llama_index/llms/ollama.py

```diff
@@ -70,38 +67,132 @@ def _model_kwargs(self) -> Dict[str, Any]:
+                    },
+                ),
+                raw=response.json(),
+                additional_kwargs={
+                    k: v for k, v in response.json().items() if k != "message"
+                },
+            )
 
     @llm_chat_callback()
     def stream_chat(
         self, messages: Sequence[ChatMessage], **kwargs: Any
     ) -> ChatResponseGen:
-        prompt = self.messages_to_prompt(messages)
-        completion_response = self.stream_complete(
-            prompt, prompt_formatted=True, **kwargs
-        )
-        return stream_completion_response_to_chat_response(completion_response)
+        try:
+            import requests
+        except ImportError:
+            raise ImportError(
+                "Could not import requests library."
+                "Please install requests with `pip install requests`"
+            )
+
+        with requests.post(
+            url=f"{self.base_url}/api/chat/",
+            json={
+                "model": self.model,
+                "messages": [
+                    {
+                        "role": message.role,
+                        "content": message.content,
+                        **message.additional_kwargs,
+                    }
+                    for message in messages
+                ],
+                "options": self._model_kwargs,
+                "stream": True,
+                **kwargs,
+            },
+            stream=True,
+        ) as response:
+            response.raise_for_status()
+            text = ""
+            for line in response.iter_lines(decode_unicode=True):
+                if line:
+                    # Parsing each line (JSON chunk) and extracting the details
+                    chunk = json.loads(line)
+                    message = chunk["message"]
+                    delta = message.get("content", "")
+                    text += delta
+                    yield ChatResponse(
+                        message=ChatMessage(
+                            message=text,
+                            role=MessageRole(message.get("role")),
+                            additional_kwargs={
+                                k: v
+                                for k, v in chunk["message"].items()
+                                if k != "content" and k != "role"
+                            },
+                        ),
+                        delta=delta,
+                        raw=chunk,
+                        additional_kwargs={
+                            k: v for k, v in chunk.items() if k != "message"
+                        },
+                    )
 
     @llm_completion_callback()
     def complete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
-        response_gen = self.stream_complete(prompt, **kwargs)
-        response_list = list(response_gen)
-        final_response = response_list[-1]
-        final_response.delta = None
-        return final_response
+        try:
```

Comment:
Does ollama handle formatting inputs for models now? We don't need `completion_to_prompt` for complete or stream complete?

Reply:
It will handle formatting inputs for models. There is a raw parameter option in case we want to format manually or overriding system prompt in Ollama Modelfile: https://github.com/jmorganca/ollama/blob/main/docs/api.md#parameters

Reply:
amazing!